logs/logs.go

```diff
@@ -0,0 +1,87 @@
+package logs
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/deis/logger/storage"
+)
+
+const (
+	controllerPattern = `^(INFO|WARN|DEBUG|ERROR)\s+(\[(\S+)\])+:(.*)`
+)
+
+//{"log"=>"2016/05/31 01:34:43 10.164.1.1 GET / - 5074209722772702441\n", "stream"=>"stderr", "docker"=>{"container_id"=>"6a9069435788a05531ee2b9afbcdc73a22018af595f3203cb67e06f50103bf5f"}, "kubernetes"=>{"namespace_name"=>"foo", "pod_id"=>"34ebc234-2423-11e6-94aa-42010a800021", "pod_name"=>"foo-v2-web-2ggow", "container_name"=>"foo-web", "labels"=>{"app"=>"foo", "heritage"=>"deis", "type"=>"web", "version"=>"v2"}, "host"=>"gke-jchauncey-default-pool-7ae1c279-10ye"}}
+
+// Message json log message from fluentd
+type Message struct {
+	Log        string     `json:"log"`
+	Stream     string     `json:"stream"`
+	Kubernetes Kubernetes `json:"kubernetes"`
+	Docker     Docker     `json:"docker"`
+}
+
+// Kuberentes decorated log message
+type Kubernetes struct {
+	Namespace     string            `json:"namespace_name"`
+	PodID         string            `json:"pod_id"`
+	PodName       string            `json:"pod_name"`
+	ContainerName string            `json:"container_name"`
+	Labels        map[string]string `json:"labels"`
+	Host          string            `json:"host"`
+}
+
+// Docker decorated log messaage
+type Docker struct {
+	ContainerID string `json:"container_id"`
+}
+
+// ReadLogs returns a specified number of log lines (if available) for a specified app by
+// delegating to the server's underlying storage.Adapter.
+func ReadLogs(storageAdapter storage.Adapter, app string, lines int) ([]string, error) {
```

Comment:
imo, it's a little ugly to expect anyone invoking the exported functions of this package to pass in a `storage.Adapter` with that function call, especially since the function is _only_ just turning around an invoking a function on the `storage.Adapter` itself (which is something that the caller of this function surely could have done themselves).

imo, the "right" way to do this is to write object oriented code. Instead of just having a `logs` package, have a log struct that holds a reference to things like a `storage.Adapter`. Allow a `storage.Adapter` to be dependency-injected when such  struct is created. Have that struct implement a public interface so that anyone invoking its methods just doesn't need to deal with the underlying storage at all.

You could probably accomplish most of that my refactoring the old `syslogish.Server` to suit your needs (or at least using it as a template).